blockmanager.go

```diff
@@ -2006,13 +2148,367 @@ out:
+// workmanager while fetching headers within the block header checkpoint region. This process is carried out
+// in order.
+func (b *blockManager) processBlKHeaderInCheckPtRegionInOrder() {
+	lenCheckPts := len(b.cfg.ChainParams.Checkpoints)
+
+	// Loop should run as long as we are in the block header checkpointed region.
+	b.newHeadersMtx.RLock()
+	for int32(b.headerTip) <= b.cfg.ChainParams.Checkpoints[lenCheckPts-1].Height {
+		hdrTip := b.headerTip
+		b.newHeadersMtx.RUnlock()
+
+		select {
+		// return quickly if the blockmanager quits.
+		case <-b.quit:
+			return
+		default:
+		}
+
+		// do not go further if we have not received the response mapped to our header tip.
+		b.writeBatchMtx.RLock()
+		msg, ok := b.hdrTipToResponse[int32(hdrTip)]
+		b.writeBatchMtx.RUnlock()
+
+		if !ok {
+			b.newHeadersMtx.RLock()
+			continue
+		}
+
+		b.syncPeerMutex.Lock()
+		b.syncPeer = msg.peer
+		b.syncPeerMutex.Unlock()
+
+		b.handleHeadersMsg(msg)
+		err := b.resetHeaderListToChainTip()
+		if err != nil {
+			log.Errorf(err.Error())
+		}
+
+		finalNode := b.headerList.Back()
+		newHdrTip := finalNode.Height
+		newHdrTipHash := finalNode.Header.BlockHash()
+		prevCheckPt := b.findPreviousHeaderCheckpoint(newHdrTip)
+
+		log.Tracef("New headertip %v", newHdrTip)
+		log.Debugf("New headertip %v", newHdrTip)
+
+		// If our header tip has not increased, there is a problem with the headers we received and so we
+		// delete all the header response within our previous header tip and our new header tip, then send
+		// another query to the workmanager.
+		if uint32(newHdrTip) <= hdrTip {
+			b.deleteHeaderTipResp(newHdrTip, int32(hdrTip))
+
+			log.Tracef("while fetching checkpointed headers received invalid headers")
+
+			q := CheckpointedBlockHeadersQuery{
+				blockMgr: b,
+				msgs: []*headerQuery{
+
+					{
+						message: &wire.MsgGetHeaders{
+							BlockLocatorHashes: []*chainhash.Hash{&newHdrTipHash},
+							HashStop:           *b.nextCheckpoint.Hash,
+						},
+						startHeight:   newHdrTip,
+						initialHeight: prevCheckPt.Height,
+						startHash:     newHdrTipHash,
+						endHeight:     b.nextCheckpoint.Height,
+						initialHash:   newHdrTipHash,
+						// Set it as high priority so that workmanager can schedule before any other queries.
+						index: 0,
+					},
+				},
+			}
+
+			b.cfg.blkHdrCheckptQueryDispatcher.Query(
+				q.requests(), query.Cancel(b.quit), query.Timeout(1*time.Hour), query.NoRetryMax(),
+			)
+
+			log.Tracef("Sending query to workmanager from processBlKHeaderInCheckPtRegionInOrder loop")
+		}
+		b.newHeadersMtx.RLock()
+	}
+	b.newHeadersMtx.RUnlock()
+
+	b.syncPeerMutex.Lock()
+	b.syncPeer = nil
+	b.syncPeerMutex.Unlock()
+
+	log.Infof("Successfully completed fetching checkpointed block headers")
+}
+
+// batchCheckpointedBlkHeaders creates headerQuery to fetch block headers
+// within the chain's checkpointed region.
+func (b *blockManager) batchCheckpointedBlkHeaders() {
```

Comment:
Hmm, I did not know about this rule.

Reply:
It's not a hard rule, especially not in the lnd/neutrino ecosystem. However, I push this direction because functions that accept no arguments and return no results necessarily cannot be useful unless all of the state they read and write is on the blockManager itself. This has a few problems:

1. it is harder to understand what a function does from glancing at its signature
2. it requires you to have a full blockManager available to be able to test it
3. by giving the function access to all of the blockManager state, it makes the boundaries between the state it needs and the state it doesn't very blurry

Reply:
> The exception to this rule is if you need to start a coroutine that is expected to run indefinitely.

Then this function is an exception to this rule, in that case. It is started as a goroutine just like `cfhandler` which has no argument and does not return any as well.


https://github.com/Chinwendu20/neutrino/blob/64b278771ff75da0b30136af7d4ab0ace06d897e/blockmanager.go#L333-L354

Reply:
> It's not a hard rule, especially not in the lnd/neutrino ecosystem. However, I push this direction because functions that accept no arguments and return no results necessarily cannot be useful unless all of the state they read and write is on the blockManager itself. This has a few problems:
> 
> 1. it is harder to understand what a function does from glancing at its signature
> 2. it requires you to have a full blockManager available to be able to test it
> 3. by giving the function access to all of the blockManager state, it makes the boundaries between the state it needs and the state it doesn't very blurry

Hmm I am thinking through this, point 2 and 3, isn't that characteristic of all methods of a struct any ways.